mixnet/mixnet.py

```diff
@@ -12,31 +16,105 @@
 class Mixnet:
     mix_nodes: List[MixNode]
 
-    # Build a new topology deterministically using an entropy.
-    # The entropy is expected to be injected from outside.
-    #
-    # TODO: Implement constructing a new topology in advance to minimize the topology transition time.
-    #       https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2?pvs=4#9a7f6089e210454bb11fe1c10fceff68
-    def build_topology(
+    def __init__(
         self,
-        entropy: bytes,
+        mix_nodes: List[MixNode],
         n_layers: int,
         n_nodes_per_layer: int,
+        mixnet_epoch_sec: int,
+        proactive_sec: int,
+        initial_entropy: bytes,
+    ):
+        self.mix_nodes = mix_nodes
+
+        self.set_entropy(initial_entropy)
+
+        self.topology_updater = MixnetTopologyUpdater(
+            mix_nodes, n_layers, n_nodes_per_layer, mixnet_epoch_sec, proactive_sec
+        )
+        self.topology_updater.daemon = True
+        self.topology_updater.start()
+
+    def set_entropy(self, entropy: bytes) -> None:
+        """
+        Set entropy (seed) to the Fisher-Yates shuffler
+        that is going to be used for deterministic topology construction.
+
+        This method is expected to be called periodically by the user.
+        Until the next entropy is provides, Fisher-Yates shuffler continues to works
+        based on this entropy set once.
+        """
+        FisherYates.set_seed(entropy)
+
+    def choose_mixnode(self) -> MixNode:
+        return random.choice(self.mix_nodes)
+
+    def current_topology(self) -> MixnetTopology:
+        return self.topology_updater.current_topology()
+
+
+class MixnetTopologyUpdater(Thread):
+    def __init__(
+        self,
+        mix_nodes: List[MixNode],
+        n_layers: int,
+        n_nodes_per_layer: int,
+        mixnet_epoch_sec: int,
+        proactive_sec: int,
+    ):
+        super().__init__()
+        self.mix_nodes = mix_nodes
+        self.n_layers = n_layers
+        self.n_nodes_per_layer = n_nodes_per_layer
+        self.mixnet_epoch_sec = mixnet_epoch_sec
+        self.proactive_sec = proactive_sec
+
+        self.lock = threading.Lock()
+        self.topology = self.build_topology()
+        self.next_topology = None
+
+    def run(self) -> None:
+        next_epoch_ts = datetime.now() + timedelta(seconds=self.mixnet_epoch_sec)
+
+        while True:
+            time.sleep(1 / 1000)
```

Comment:
Why is there a sleep here? nitpick: I think schedules are defined already when checking it below. Probably can be abstracted into some scheduler. 
Not necessary to change now. But we could have a blackboxed structure than awaits till next epoch.